### ISSUE TITLE
Function availability must be tested against build-time cuSOLVER version, not against build-time CUDA Runtime version

### ISSUE THREAD

**Issue by kmaehashi** [cupy/cupy#5530]
Currently each library in CUDA Toolkit is versioned independently of the CUDA Toolkit versions, so we should test against the library version. We should also update `#ifdef` s in stub too.

https://github.com/cupy/cupy/blob/d3c46cfdba25c4d0892b76a4f89e757a3bd05e18/cupy/cusolver.pyx#L93

**Comment by leofang** [comment]
Maybe I misunderstood, wouldn't this conflict with the idea of #5401?

**Comment by kmaehashi** [comment]
No, these two don't conflict. Currently CuPy uses `_driver.get_build_version()` (i.e., `CUDA_VERSION`, build-time CUDA Runtime version) to test availability, but this should be `cusolver.get_build_version()` (which is not available at this time).

This is because `CUDA_VERSION` does not change between patch releases (e.g., CUDA 11.3.0 and 11.3.1), but cuSPARSE's version may be different between two:

https://gitlab.com/nvidia/headers/cuda/-/commit/7f54cb8e9f5ff072946e099e60d701d65ef03c7f

This is why we are using `cusparse.get_build_version()` in cuSPARSE.

https://github.com/cupy/cupy/blob/d3c46cfdba25c4d0892b76a4f89e757a3bd05e18/cupy/cusparse.py#L159

Luckily, this wasn't the issue because cuSOLVER only changes the patch versions between CUDA patch releases, but  that's not guaranteed.

**Comment by leofang** [comment]
Ah I see, thanks for clarification!

**Comment by kmaehashi** [comment]
This will be a mandatory fix if we rewrite cuSOLVER with CUDA Python.
